### PR TITLE
Adding interfaces and base classes for exceptions

### PR DESCRIPTION
**Pull request by dsgouda**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] Please add REST spec PR link to the SDK PR
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

### CODE REVIEW

File: src/SdkCommon/ClientRuntime/ClientRuntime/IRestException.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+// Licensed under the MIT License. See License.txt in the project root for license information.
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Rest
+{
+    public interface IRestException
```
Comment (shahabhijeet) [inline src/SdkCommon/ClientRuntime/ClientRuntime/IRestException.cs]: @dsgouda something to consider this for this feature.
Your intend is to provide a constraint for your generated exceptions and not create exceptions and return for clients to use it.
Move properties out of the Interface and into the base class

File: src/SdkCommon/ClientRuntime/ClientRuntime/RestException.cs [hunk 2]
```diff
@@ -12,7 +12,7 @@ namespace Microsoft.Rest
 #if FullNetFx
     [Serializable]
 #endif
-    public class RestException : Exception
+    public class RestException : Exception, IRestException
```
Comment (shahabhijeet) [inline src/SdkCommon/ClientRuntime/ClientRuntime/RestException.cs]: @dsgouda do not change existing RestException.
This change can be made during breaking change cycle.

File: src/SdkCommon/ClientRuntime/ClientRuntime/RestExceptionGeneric.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        /// <param name="requestMessage">The request message.</param>
+        /// <param name="responseMessage">The response message.</param>
+        public RestException(string message, HttpRequestMessageWrapper requestMessage, HttpResponseMessageWrapper responseMessage, int statusCode) : base(message)
+        {
+            Request = requestMessage;
+            Response = responseMessage;
+            ResponseStatusCode = statusCode;
+        }
+
+        public V Body { get; private set; }
```
Comment (shahabhijeet) [inline src/SdkCommon/ClientRuntime/ClientRuntime/RestExceptionGeneric.cs]: Change Body to ErrorModel or ResponseErrorModel. Bascially idea is to use the same name that we use during swagger review.
Remove private setter (basically nothing that you gain out of making it private) if made public also aligns with CloudException body property

File: src/SdkCommon/ClientRuntime/ClientRuntime/RestExceptionGeneric.cs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        /// Gets information about the associated HTTP response.
+        /// </summary>
+        public HttpResponseMessageWrapper Response { get; set; }
+
+        /// <summary>
+        /// the status code returned by server
+        /// </summary>
+        public int ResponseStatusCode { get; set; }
+
+        public virtual void SetErrorModel(V model)
```
Comment (shahabhijeet) [inline src/SdkCommon/ClientRuntime/ClientRuntime/RestExceptionGeneric.cs]: @dsgouda once Body/ErrorModel has public setter, this function is no longer needed.

File: src/SdkCommon/ClientRuntime/ClientRuntime/RestExceptionGeneric.cs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+        /// <summary>
+        /// Gets information about the associated HTTP response.
+        /// </summary>
+        public HttpResponseMessageWrapper Response { get; set; }
+
+        /// <summary>
+        /// the status code returned by server
+        /// </summary>
+        public int ResponseStatusCode { get; set; }
```
Comment (shahabhijeet) [inline src/SdkCommon/ClientRuntime/ClientRuntime/RestExceptionGeneric.cs]: @dsgouda we no longer need this as discussed during PR review.

**Review by shahabhijeet — CHANGES_REQUESTED** [review]

File: src/SdkCommon/ClientRuntime/ClientRuntime/RestExceptionGeneric.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+#if FullNetFx
+    [Serializable]
+#endif
+    public class RestException<T> : RestException
+    {
+        
+        /// <summary>
+        /// Initializes a new instance of the ErrorModelException class.
+        /// </summary>
+        public RestException()
```
Comment (dsgouda) [inline src/SdkCommon/ClientRuntime/ClientRuntime/RestExceptionGeneric.cs]: Call the base from constructor

File: src/SdkCommon/ClientRuntime/ClientRuntime/RestExceptionGeneric.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the CloudException class caused by another exception.
+        /// </summary>
+        /// <param name="message">A description of the error.</param>
+        /// <param name="requestMessage">The request message.</param>
+        /// <param name="responseMessage">The response message.</param>
+        public RestException(string message, HttpRequestMessageWrapper requestMessage, HttpResponseMessageWrapper responseMessage) : base(message)
```
Comment (dsgouda) [inline src/SdkCommon/ClientRuntime/ClientRuntime/RestExceptionGeneric.cs]: Remove this

File: src/SdkCommon/ClientRuntime/ClientRuntime/IRestErrorModel.cs [hunk 8]
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
```
Comment (shahabhijeet) [inline src/SdkCommon/ClientRuntime/ClientRuntime/IRestErrorModel.cs]: missing xml comment avoids compiler warnings (we are already have more than 1000 warnings in our repo)
Please describe the interface and the purpose of the interface.

File: src/SdkCommon/ClientRuntime/ClientRuntime/RestExceptionGeneric.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
```
Comment (shahabhijeet) [inline src/SdkCommon/ClientRuntime/ClientRuntime/RestExceptionGeneric.cs]: xml comments and description for the class.
All public APIs get xml comments

File: src/SdkCommon/ClientRuntime/ClientRuntime/RestExceptionGeneric.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        /// <summary>
+        /// Gets information about the associated HTTP request.
+        /// </summary>
+        public HttpRequestMessageWrapper Request { get; set; }
+
+        /// <summary>
+        /// Gets information about the associated HTTP response.
+        /// </summary>
+        public HttpResponseMessageWrapper Response { get; set; }
+
```
Comment (shahabhijeet) [inline src/SdkCommon/ClientRuntime/ClientRuntime/RestExceptionGeneric.cs]: missing RequestId as we discussed earlier

**Review by shahabhijeet — CHANGES_REQUESTED** [review]

**Pull request closed without merging**